extensions/guacamole-auth-totp/src/main/java/org/apache/guacamole/auth/totp/user/CodeUsageTrackingService.java

```diff
@@ -0,0 +1,264 @@
+        executor.scheduleAtFixedRate(new CodeEvictionTask(), 1, 1, TimeUnit.MINUTES);
+    }
+
+    /**
+     * Task which iterates through all explicitly-invalidated codes, evicting
+     * those codes which are old enough that they would fail validation against
+     * the secret key anyway.
+     */
+    private class CodeEvictionTask implements Runnable {
+
+        @Override
+        public void run() {
+
+            // Get start time of cleanup check
+            long checkStart = System.currentTimeMillis();
+
+            // For each code still being tracked, remove those which are old
+            // enough that they would fail validation against the secret key
+            Iterator<Map.Entry<UsedCode, Long>> entries = invalidCodes.entrySet().iterator();
+            while (entries.hasNext()) {
+
+                Map.Entry<UsedCode, Long> entry = entries.next();
+                long invalidUntil = entry.getValue();
+
+                // If code is sufficiently old, evict it and check the next one
+                if (checkStart >= invalidUntil)
+                    entries.remove();
+
+            }
+
+            // Log completion and duration
+            logger.debug("TOTP tracking cleanup check completed in {} ms.",
+                    System.currentTimeMillis() - checkStart);
+
+        }
+
+    }
+
+    /**
+     * A valid TOTP code which was previously used by a particular user.
+     */
+    private class UsedCode {
+
+        /**
+         * The username of the user which previously used this code.
+         */
+        private final String username;
+
+        /**
+         * The valid code given by the user.
+         */
+        private final String code;
+
+        /**
+         * Creates a new UsedCode which records the given code as having been
+         * used by the given user.
+         *
+         * @param username
+         *     The username of the user which previously used the given code.
+         *
+         * @param code
+         *     The valid code given by the user.
+         */
+        public UsedCode(String username, String code) {
+            this.username = username;
+            this.code = code;
+        }
+
+        /**
+         * Returns the username of the user which previously used the code
+         * associated with this UsedCode.
+         *
+         * @return
+         *     The username of the user which previously used this code.
+         */
+        public String getUsername() {
+            return username;
+        }
+
+        /**
+         * Returns the valid code given by the user when this UsedCode was
+         * created.
+         *
+         * @return
+         *     The valid code given by the user.
+         */
+        public String getCode() {
+            return code;
+        }
+
+        @Override
+        public int hashCode() {
+            int hash = 7;
+            hash = 79 * hash + this.username.hashCode();
+            hash = 79 * hash + this.code.hashCode();
```

Comment:
Maybe this is part of the IETF reference implementation, but is there some significance to the values 7 and 79, here?  Is it worth making these constants, or is there a particular reason they should not be made constant?

Reply:
> Maybe this is part of the IETF reference implementation, but is there some significance to the values 7 and 79, here?

Nope, it's not specific to the IETF implementation. It's a generic hash function for the values within this class. The primes 7 and 79 were arbitrarily chosen and have no meaning other than to generally distribute the values well.

Reply:
ok